vm/src/stdlib/operator.rs

```diff
@@ -1,62 +1,731 @@
+    // a.__index__ as raised in the issue. Currently, we accept int subclasses.
+    /// Return a converted to an integer. Equivalent to a.__index__().
+    #[pyfunction]
+    fn index(a: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyIntRef> {
+        vm.to_index(&a)
+    }
+
+    /// Return the bitwise inverse of the number obj. This is equivalent to ~obj.
+    #[pyfunction]
+    fn invert(pos: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._invert(&pos)
+    }
+
+    /// Return a shifted left by b.
+    #[pyfunction]
+    fn lshift(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._lshift(&a, &b)
+    }
+
+    /// Return a % b
+    #[pyfunction(name = "mod")]
+    fn mod_(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._mod(&a, &b)
+    }
+
+    /// Return a * b
+    #[pyfunction]
+    fn mul(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._mul(&a, &b)
+    }
+
+    /// Return a @ b
+    #[pyfunction]
+    fn mat_mul(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._matmul(&a, &b)
+    }
+
+    /// Return obj negated (-obj).
+    #[pyfunction]
+    fn neg(pos: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._neg(&pos)
+    }
+
+    /// Return the bitwise or of a and b.
+    #[pyfunction]
+    fn or_(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._or(&a, &b)
+    }
+
+    /// Return obj positive (+obj).
+    #[pyfunction]
+    fn pos(obj: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._pos(&obj)
+    }
+
+    /// Return a ** b, for a and b numbers.
+    #[pyfunction]
+    fn pow(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._pow(&a, &b)
+    }
+
+    /// Return a shifted right by b.
+    #[pyfunction]
+    fn rshift(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._rshift(&a, &b)
+    }
+
+    /// Return a - b.
+    #[pyfunction]
+    fn sub(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._sub(&a, &b)
+    }
+
+    /// Return a / b where 2/3 is .66 rather than 0. This is also known as "true" division.
+    #[pyfunction]
+    fn truediv(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._truediv(&a, &b)
+    }
+
+    /// Return the bitwise exclusive or of a and b.
+    #[pyfunction]
+    fn xor(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        vm._xor(&a, &b)
+    }
+
+    // Sequence based operators
+
+    /// Return a + b for a and b sequences.
+    #[pyfunction]
+    fn concat(a: PyObjectRef, b: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        // Best attempt at checking that a is sequence-like.
```

Comment:
Looking for tips on how to improve this (and `iconcat`) if anyone has ideas!

Reply:
Maybe PySequence_Check in CPython? https://docs.python.org/3/c-api/sequence.html

I have no idea in RustPython.

Reply:
Yeah, it's similar to the question I asked you the other day about a `PyNumber_Check` like method. We don't currently have the abstract interface.